edgedb/asyncio_con.py

```diff
@@ -32,41 +35,160 @@
+        return result
+
+
+class _AsyncIOConnectionImpl:
+
+    def __init__(self, codecs_registry, query_cache):
+        self._codecs_registry = codecs_registry
+        self._query_cache = query_cache
+
+        # TODO(tailhook) temporary, this is not needed when we implement
+        # transaction part of RFC1004
+        self._top_xact = None
 
-    def __instancecheck__(cls, instance):
-        mro = type(instance).__mro__
-        return AsyncIOConnection in mro or _ConnectionProxy in mro
+    def is_closed(self):
+        transport = self._transport
+        protocol = self._protocol
+        return not (
+            transport and not transport.is_closing() and
+            protocol and protocol.connected
+        )
+
+    async def connect(self, loop, addrs, config, params):
+        addr = None
+        max_time = time.monotonic() + config.wait_until_available
+        iteration = 1
+
+        while True:
+            for addr in addrs:
+                try:
+                    con = await asyncio.wait_for(
+                        self._connect_addr(loop, addr, params),
+                        config.connect_timeout,
+                    )
+                except asyncio.TimeoutError as e:
+                    if iteration == 1 or time.monotonic() < max_time:
+                        continue
+                    else:
+                        raise errors.ConnectionTimeoutError(
+                            f"connecting to {addr} failed in"
+                            f" {config.connect_timeout} sec"
+                        ) from e
+                except errors.ClientConnectionError as e:
+                    if (
+                        e.has_tag(errors.SHOULD_RECONNECT) and
+                        (iteration == 1 or time.monotonic() < max_time)
+                    ):
+                        continue
+                    e.message = con_utils.render_client_no_connection_error(
+                        e,
+                        addr,
+                    )
+                    raise e
+                else:
+                    self._addr = addr
+                    return con
+
+            iteration += 1
+            await asyncio.sleep(0.01 + random.random() * 0.2)
+
+    async def _connect_addr(self, loop, addr, params):
+
+        factory = lambda: asyncio_proto.AsyncIOProtocol(
+            params, loop)
+
+        try:
+            if isinstance(addr, str):
+                # UNIX socket
+                tr, pr = await loop.create_unix_connection(factory, addr)
+            else:
+                tr, pr = await loop.create_connection(factory, *addr)
+        except socket.gaierror as e:
+            # All name resolution errors are considered temporary
+            raise errors.ConnectionFailedTemporarilyError(str(e)) from e
+        except OSError as e:
+            # TODO(tailhook) figure out error
+            message = str(e)
+            if e.errno is None:
+                errnos = extract_errno(message)
```

Comment:
Why not `e.errno`?

Reply:
Because it's absent in "Multiple exceptions" case. I've added a docstring in `_extract_errno` text